### ISSUE TITLE
cstor-pool container unable to start due to libjson-c.so.2

### ISSUE THREAD

**Issue by xUnholy** [openebs/openebs#3037]
# Description

I got some strange errors about the libjson library. Apologies if this is in the wrong repository.

❯ k logs cstor-sparse-pool-w39i-65b4d88bbf-f4mzd cstor-pool
Disabling dumping core
sleeping for 2 sec
/usr/local/bin/zrepl: error while loading shared libraries: libjson-c.so.2: cannot open shared object file: No such file or directory

**Environment:**
- OpenEBS version: 1.9.0 (Also tried using 1.10.x.RC2)
- Kubernetes version: 1.18.x
- OS: Ubuntu 20.04 Focal
- Hardware: Raspberry Pi 4B

**Comment by kmova** [comment]
Hi @xUnholy -- this is the right repository. Thanks for reporting this issue. I don't think we tried to test with Ubuntu 20.04 out yet. 

@mynktl - may need some further information to debug this issue.

**Comment by xUnholy** [comment]
@mynktl all my other pods are starting as per normal.

I'm deploying with the helm approach, below is the values I'm using.

```
rbac:
  pspEnabled: true
ndm:
  image: 'openebs/node-disk-manager-arm64'
  sparse:
    count: '1'
ndmOperator:
  image: 'openebs/node-disk-operator-arm64'
webhook:
  image: 'openebs/admission-server-arm64'
apiserver:
  image: 'openebs/m-apiserver-arm64'
  sparse:
    enabled: true
localprovisioner:
  image: 'openebs/provisioner-localpv-arm64'
snapshotOperator:
  controller:
    image: 'openebs/snapshot-controller-arm64'
  provisioner:
    image: 'openebs/snapshot-provisioner-arm64'
provisioner:
  image: 'openebs/openebs-k8s-provisioner-arm64'
helper:
  image: 'openebs/linux-utils-arm64'
cstor:
  pool:
    image: 'openebs/cstor-pool-arm64'
  poolMgmt:
    image: 'openebs/cstor-pool-mgmt-arm64'
  target:
    image: 'openebs/cstor-istgt-arm64'
  volumeMgmt:
    image: 'openebs/cstor-volume-mgmt-arm64'
policies:
  monitoring:
    image: 'openebs/m-exporter-arm64'
```

I'm running on a fresh Ubuntu 20.04 Focal OS, I followed the pre-reqs and confirmed iscsid is working fine.

```
● iscsid.service - iSCSI initiator daemon (iscsid)
     Loaded: loaded (/lib/systemd/system/iscsid.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2020-05-12 14:22:59 AEST; 2 days ago
TriggeredBy: ● iscsid.socket
       Docs: man:iscsid(8)
    Process: 736298 ExecStartPre=/lib/open-iscsi/startup-checks.sh (code=exited, status=0/SUCCESS)
    Process: 736303 ExecStart=/sbin/iscsid (code=exited, status=0/SUCCESS)
   Main PID: 736305 (iscsid)
      Tasks: 2 (limit: 4434)
     Memory: 3.2M
     CGroup: /system.slice/iscsid.service
             ├─736304 /sbin/iscsid
             └─736305 /sbin/iscsid
```

Upon seeing the error I also `sudo apt install -y libjson-c-dev` but to no avail. I'm creating the StorPool WITHOUT specifying diskList and it appears to be fine.

**Comment by sgielen** [comment]
I had this issue as well. It is caused by the fact that zrepl is built outside Docker, so it depends on the libraries available on the building host (in this case libjson-c.so.2), and then the binary is copied into a Docker image where other libraries are available (in this case libjson-c.so.3).

I fixed it using this Dockerfile, which takes cstor-pool-arm64:1.9.0 and installs libjson-c.so.2 on it:

```
FROM ubuntu:bionic AS build

RUN apt-get update && apt-get install -y build-essential wget devscripts debhelper dh-exec dh-autoreconf

WORKDIR /root

RUN wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11-4ubuntu2.dsc && \
	wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11.orig.tar.gz && \
	wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11-4ubuntu2.debian.tar.xz && \
	dpkg-source -x json-c_0.11-4ubuntu2.dsc && \
	cd json-c-0.11 && \
	DEB_CFLAGS_SET="-Wno-error=implicit-fallthrough" debuild -us -uc

FROM quay.io/openebs/cstor-pool-arm64:1.9.0 AS run

COPY --from=build /root/libjson-c2_0.11-4ubuntu2_arm64.deb /tmp/libjson-c2.deb
RUN dpkg -i /tmp/libjson-c2.deb && rm /tmp/libjson-c2.deb
```

**Comment by xUnholy** [comment]
Thanks for your input @sgielen - Did you manage to publish your image that could be reused rather than myself building it again? This should be good feedback to the upstream.

**Comment by sgielen** [comment]
> Thanks for your input @sgielen - Did you manage to publish your image that could be reused rather than myself building it again? This should be good feedback to the upstream.

I have it on a private repo, sorry, don't currently have a public repository running to make it available...

**Comment by xUnholy** [comment]
@sgielen I'm trying to build your Dockerfile although there appear to be some issues - https://github.com/raspbernetes/multi-arch-images/runs/674171607

**Comment by sgielen** [comment]
> @sgielen I'm trying to build your Dockerfile although there appear to be some issues - https://github.com/raspbernetes/multi-arch-images/runs/674171607
> 
> Looks like `COPY --from=build /root/libjson-c2_0.11-4ubuntu2_arm64.deb /tmp/libjson-c2.deb` doesn't exist in the build stage, you built this lately?

That one is successfully built just before in the earlier stage of the multistage build:

```
dpkg-deb: building package 'libjson-c2' in '../libjson-c2_0.11-4ubuntu2_amd64.deb'.
```

Since the workdir is `/root`, the `.deb` is correctly written according to the output and the `COPY --from=build` fails, is it possible that the build system you are using does not support multistage builds?

**Comment by xUnholy** [comment]
It supports it but your COPY is pulling arm64 whereas you're downloading amd64, was this intended?

**Comment by sgielen** [comment]
> It supports it but your COPY is pulling arm64 whereas you're downloading amd64, was this intended?

Ah, that's it: I don't specify the arch anywhere in the first stage of the build, but I run it on an arm64 node, therefore it buids arm64 libjson. For you, probably, the first stage ran on an amd64 machine, so it build amd64 libjson and therefore is unable to find the arm64 file.

Could you try replacing the `FROM` of the build stage with an image that is only available for an arm64 host, or in some other way force the build to run on arm64?

**Comment by mynktl** [comment]
Thanks @sgielen for helping @xUnholy .
@xUnholy Hope you are able to build the image successfully. As of now I don't have arm machine to verify the image, i'll test it tomorrow and push the fix.
Meanwhile, i'm posting the docker file here. I've done minor modification.
```
FROM ubuntu AS build

WORKDIR /root

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && apt-get install -y build-essential wget devscripts debhelper dh-exec dh-autoreconf

RUN wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11-4ubuntu2.dsc && \
  wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11.orig.tar.gz && \
  wget http://archive.ubuntu.com/ubuntu/pool/main/j/json-c/json-c_0.11-4ubuntu2.debian.tar.xz && \
  dpkg-source -x json-c_0.11-4ubuntu2.dsc && \
  cd json-c-0.11 && \
  DEB_CFLAGS_SET="-Wno-error=implicit-fallthrough" debuild -us -uc && \
  chmod +x /root/libjson-c2_0.11-4ubuntu2_arm64.deb


FROM quay.io/openebs/cstor-pool-arm64:1.9.0 AS run

COPY --from=build /root/libjson-c2_0.11-4ubuntu2_arm64.deb /tmp/libjson-c2.deb
RUN dpkg -i /tmp/libjson-c2.deb && rm /tmp/libjson-c2.deb 

```

**Comment by xUnholy** [comment]
For anyone following this thread, I've compiled a temporary image until the upstream works appropriately.

docker.io/raspbernetes/cstor-pool-arm64:1.9.0

**Comment by sgielen** [comment]
This is fixed in OpenEBS 1.11 and can be closed.
